pkg/builder/builder_test.go

```diff
@@ -114,7 +115,72 @@ func successfullyBuildDefaultStageAndFileTC(mockCtrl *gomock.Controller) testCas
 		LockFile:    "zbuild.lock",
 		Stage:       "dev",
 		SessionID:   "<SESSION-ID>",
-		ContextName: "context",
+		ContextName: "some-context-name",
+	}
+	state := llb.State{}
+	img := image.Image{
+		Image: specs.Image{
+			Author: "zbuild",
+		},
+	}
+	handler := mocks.NewMockKindHandler(mockCtrl)
+	handler.EXPECT().WithSolver(gomock.Any()).Times(1)
+	handler.EXPECT().Build(
+		ctx, MatchBuildOpts(buildOpts),
+	).Return(state, &img, nil)
+
+	registry := registry.NewKindRegistry()
+	registry.Register("php", handler)
+
+	refImage := llbtest.NewMockReference(mockCtrl)
+	resImg := &client.Result{
+		Refs: map[string]client.Reference{"linux/amd64": refImage},
+		Ref:  refImage,
+	}
+	c.EXPECT().Solve(gomock.Any(), gomock.Any()).Return(resImg, nil)
+
+	imgConfig := `{"author":"zbuild","architecture":"","os":"","rootfs":{"type":"","diff_ids":null},"config":{}}`
+	return testCase{
+		client:   c,
+		solver:   solver,
+		registry: registry,
+		expectedRes: &client.Result{
+			Refs: map[string]client.Reference{"linux/amd64": refImage},
+			Ref:  refImage,
+			Metadata: map[string][]byte{
+				"containerimage.config": []byte(imgConfig),
+			},
+		},
+	}
+}
+
+func successfullyBuildFromGitContextTC(mockCtrl *gomock.Controller) testCase {
+	c := llbtest.NewMockClient(mockCtrl)
+	c.EXPECT().BuildOpts().AnyTimes().Return(client.BuildOpts{
+		SessionID: "<SESSION-ID>",
+		Opts: map[string]string{
+			"context": "git://github.com/some/repo",
+		},
+	})
+
+	solver := mocks.NewMockStateSolver(mockCtrl)
+	solver.EXPECT().FromBuildContext(gomock.Any()).Times(1)
+
+	solver.EXPECT().ReadFile(
+		gomock.Any(), "zbuild.yml", gomock.Any(),
+	).Return(zbuildYml, nil)
+
+	solver.EXPECT().ReadFile(
+		gomock.Any(), "zbuild.lock", gomock.Any(),
+	).Return(zbuildLock, nil)
+
+	ctx := context.TODO()
+	buildOpts := builddef.BuildOpts{
+		File:        "zbuild.yml",
+		LockFile:    "zbuild.lock",
+		Stage:       "dev",
+		SessionID:   "<SESSION-ID>",
+		ContextName: "git://github.com/some/repo",
```

Comment:
will it be possible to set the app root path from the cloned repo ?
(mono repo vs multi repo apps arch)

Reply:
Good point, I didn't thought about that! I guess the app root should be the base path of the zbuildfile :thinking:

Btw I'm thinking about splitting the build context into two parts: 1. `source_context` where the source code is available and 2. `config_context` where zbuildfiles and other config files would be stored. For instance, rather than forking/cloning Taiga repo to accommodate its Docker setup to our needs, we'd only create a repo containing our custom config files, write a zbuildfile there and then build it from that repo.
However this split would not be handled well by docker CLI as it expects only a single context but maybe this could be circumvented by using a build-arg to specify the source_context.

Reply:
Btw, I'll fix the base path issue in a separate PR, I've +30 pending commits right now :smile:

Reply:
> For instance, rather than forking/cloning Taiga repo to accommodate its Docker setup to our needs, we'd only create a repo containing our custom config files, write a zbuildfile there and then build it from that repo.

So this covers the need where we want to zbuild apps which aren't ours (i.e. that we don't directly edit the source code). Is the scenario meant to happen often ?
Does it mean that this split should be done for our apps too ? (i.e. source code vs zbuild config and service config).

Does the `xxx_context` can be a path and / or a git path ? e.g. both paths for our apps, and a path for taiga zbuild config + a repo for taiga source ?

Reply:
> I guess the app root should be the base path of the zbuildfile

:+1: could work indeed

Reply:
> So this covers the need where we want to zbuild apps which aren't ours (i.e. that we don't directly edit the source code). Is the scenario meant to happen often ?

Exactly! I believe this could be useful whenever someone wants to create custom images for FOSS projects :wink:

> Does it mean that this split should be done for our apps too ? (i.e. source code vs zbuild config and service config).

Nope it wouldn't be needed as the source context would be the fallback context whenever there's no config context provided.

> Does the xxx_context can be a path and / or a git path ? e.g. both paths for our apps, and a path for taiga zbuild config + a repo for taiga source ?

Yep that's the idea, having two separate contexts, with one optional and the other acting as a fallback, and both could either be local or git-based :smile: